moving typescript files and compressing into one doc in reference (#826)
denoland/docs: runtime/reference/ts_config_migration.md

---
title: "Configuring TypeScript"
oldUrl:
  - /runtime/manual/advanced/typescript/faqs/
  - /runtime/manual/advanced/typescript/migration/
  - /runtime/manual/advanced/typescript/configuration/
  - /runtime/manual/advanced/typescript/types/
  - /runtime/manual/typescript/types/
  - /runtime/manual/advanced/faqs/
---

Deno’s flexibility shines in its equal treatment of TypeScript and JavaScript.
Whether you’re transitioning from JavaScript to TypeScript or vice versa, Deno
has features to ease the journey.

## Type Checking JavaScript

You may wish to make your JavaScript more type-sound without adding type
annotations everywhere. Deno supports using the TypeScript type checker to type
check JavaScript. You can mark individual files by adding the check JavaScript
pragma to the file:

```js
// @ts-check
```

This will cause the type checker to infer type information about the JavaScript
code and raise any issues as diagnostic issues.

These can be turned on for all JavaScript files in a program by providing a
configuration file with the check JS option enabled:

```json
{
  "compilerOptions": {
    "checkJs": true
  }
}
```

> And then setting the `--config` option on the command line.

Reviewer: Nit: This reads a bit odd as a sentence alone. It makes sense if it would be part of the previous one.

## Using JSDoc in JavaScript

When type-checking JavaScript or importing JavaScript into TypeScript, JSDoc
annotations can provide additional type information beyond what can just be
inferred from the code itself. Deno supports this seamlessly if you annotate
your code inline with the supported
[TypeScript JSDoc](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html).

For example to set the type of an array use the following JSDoc comment:

```js
/** @type {string[]} */
const a = [];
```

## Skipping type checking

You might have TypeScript code that you are experimenting with, where the syntax
is valid but not fully type safe. You can bypass type checking for a whole
program by passing the `--no-check` flag.

You can also skip whole files being type checked, including JavaScript if you
have check JS enabled, by using the `nocheck` pragma:

```js
// @ts-nocheck
```

## Renaming JS files to TS files

> While this might work in some cases, it has some severe limits in Deno. This is
> because Deno, by default, runs type checking in what is called _strict mode_.
> This means a lot of unclear or ambiguous situations where are not caught in
> non-strict mode will result in diagnostics being generated, and JavaScript is
> nothing but unclear and ambiguous when it comes to types.

Reviewer: A bit worried about the `it has some sever limits in Deno` phrasing. Makes it sound like it's something a user shouldn't do. We want to encourage users to go with TS. Maybe something like:

```suggestion
TypeScript files benefit from the TypeScript compiler being able to do more thorough safety checks of your code. This is often referred to as _strict mode_. When you rename a `.js` file to `.ts` you'll might see new type errors that you TypeScript wasn't able to detect before.
```
```

## Configuring TypeScript in Deno

> While TypeScript offers many configuration options, which can be daunting if
> you're just starting out with TS. Deno aims to simplify using TypeScript,
> instead of drowning you in countless settings, Deno configures TypeScript to
> **just work** out of the box. No extra configuration headaches required!

Reviewer: The first sentence reads a bit odd:

> While TypeScript offers many configuration options, which can be daunting if
you're just starting out with TS.

Maybe we can start with the good parts:

```suggestion
TypeScript is configured out of the box as you'd expect with Deno. For most projects you'll never have to touch any TypeScript related configuration options. If you do run into the need to change some of the TypeScript's compiler options, you can do so via `deno.json`.
```

Author: Have updated

| `strictNullChecks`               | `true`                  |                                                                                                                                           |
| `suppressExcessPropertyErrors`   | `false`                 |                                                                                                                                           |
| `suppressImplicitAnyIndexErrors` | `false`                 |                                                                                                                                           |
| `useUnknownInCatchVariables`     | `false`                 |                                                                                                                                           |

For a full list of compiler options and how they affect TypeScript, please refer
to the
[TypeScript Handbook](https://www.typescriptlang.org/docs/handbook/compiler-options.html).

## Using the "lib" property

> Deno includes unique built-in libraries not found in other platforms (like
> `tsc`). This can sometimes pose challenges when writing code meant to run in
> both Deno and a browser. In such cases, you can tweak Deno’s behavior using the
> "lib" property within your `compilerOptions`.

Reviewer: Nit: There shouldn't be "challenges" with Deno when it comes to the wording. Maybe something like:

```suggestion
If you're working on a project that ships code to multiple runtimes, like browsers for example, you can tweak the default types via the "lib" property within the `compilerOptions`.
```

The built-in libraries that are of interest to users:

- `"deno.ns"` - This includes all the custom `Deno` global namespace APIs plus
  the Deno additions to `import.meta`. This should generally not conflict with
  other libraries or global types.
- `"deno.unstable"` - This includes the addition unstable `Deno` global
  namespace APIs.
- `"deno.window"` - This is the "default" library used when checking Deno main
  runtime scripts. It includes the `"deno.ns"` as well as other type libraries
  for the extensions that are built into Deno. This library will conflict with
  libraries like `"dom"` and `"dom.iterable"` that are standard TypeScript
  libraries.
- `"deno.worker"` - This is the library used when checking a Deno web worker
  script. For more information about web workers, check out
  [Type Checking Web Workers](./types.md#type-checking-web-workers).
- `"dom.asynciterable"` - TypeScript currently does not include the DOM async
  iterables that Deno implements (plus several browsers), so we have implemented
  it ourselves until it becomes available in TypeScript.

> These are common libraries that Deno doesn't use, but are useful when writing
> code that is intended to also work in another runtime:

Reviewer: ```suggestion
These are common libraries that are not enabled by default, but are useful when writing
code that is intended to also work in another runtime:
```

using any APIs exclusive to one or the other. In such cases, a typical
`compilerOptions` configuration might look like this:

```json title="deno.json"
{
  "compilerOptions": {
    "target": "esnext",
    "lib": ["dom", "dom.iterable", "dom.asynciterable", "deno.ns"]
  }
}
```

This should allow most code to be type checked properly by Deno.

If you expect to run the code in Deno with the `--unstable` flag, then you
should add that library to the mix as well:

```json title="deno.json"
{
  "compilerOptions": {
    "target": "esnext",
    "lib": [
      "dom",
      "dom.iterable",
      "dom.asynciterable",
      "deno.ns",
      "deno.unstable"
    ]
  }
}
```

Typically, when you use the `"lib"` option in TypeScript, you need to include an
"es" library as well. In the case of `"deno.ns"` and `"deno.unstable"`, they
automatically include `"esnext"` when you bring them in.

The biggest "warning" when doing something like this, is that the type checking
is significantly looser, and there is no way to validate that you are doing
sufficient and effective feature detection in your code, which may lead to what

> could be trivial errors becoming runtime errors.

Reviewer: Not sure what this is referring to. I might be missing context.

Author: lets just take it out, i don't think its particularly helpful text anyway

:::note

> If you get type errors like `**cannot find`document`or`HTMLElement`**, it is

Reviewer: ```suggestion
If you get type errors like `**cannot find `document` or `HTMLElement`**, it is
```

to apply the standard Deno library as well. The second instructs Deno to apply
the built-in Deno worker type definitions plus dependent libraries (like
`"esnext"`).

When you run a `deno cache` or `deno bundle` command or use an IDE which uses
the Deno language server, Deno should automatically detect these directives and
apply the correct libraries when type checking.

The one disadvantage of this, is that it makes the code less portable to other
non-Deno platforms like `tsc`, as it is only Deno which has the `"deno.worker"`
library built into it.

### Providing "lib" setting in deno.json

You can provide a "lib" option in your `deno.json` file to instruct Deno to use
library files. For example:

```json title="deno.json"
{
  "compilerOptions": {
    "target": "esnext",
    "lib": ["deno.worker"]
  }
}
```

Then when running deno subcommand, you would need to pass the
`--config path/to/file` argument, or if you are using an IDE which leverages the
Deno language server, set the `deno.config` setting.

If you also have non-worker scripts, you will either need to omit the `--config`
argument, or have one that is configured to meet the needs of your non-worker

> scripts.

Reviewer: This seems to duplicate an earlier section about supporting the browser.

Author: I wasn't sure whether it was worth reiterating that it works for workers too, think we should take it out?
